pilot/pkg/xds/proxy_dependencies.go

```diff
@@ -89,24 +108,19 @@ func proxyDependentOnConfig(proxy *model.Proxy, config model.ConfigKey, push *mo
 	return false
 }
 
-// DefaultProxyNeedsPush check if a proxy needs push for this push event.
-func DefaultProxyNeedsPush(proxy *model.Proxy, req *model.PushRequest) bool {
-	if ConfigAffectsProxy(req, proxy) {
-		return true
+// DefaultProxyNeedsPush check if a proxy needs push for this push event and returns a new PushRequest in the case
+// it needs to filter relevant updates for this proxy.
+func DefaultProxyNeedsPush(proxy *model.Proxy, req *model.PushRequest) (*model.PushRequest, bool) {
+	if req.Forced {
+		return req, true
 	}
 
-	// If the proxy's service updated, need push for it.
-	if len(proxy.ServiceTargets) > 0 && req.ConfigsUpdated != nil {
-		for _, svc := range proxy.ServiceTargets {
-			if _, ok := req.ConfigsUpdated[model.ConfigKey{
-				Kind:      kind.ServiceEntry,
-				Name:      string(svc.Service.Hostname),
-				Namespace: svc.Service.Attributes.Namespace,
-			}]; ok {
-				return true
-			}
-		}
+	if proxy.IsWaypointProxy() || proxy.IsZTunnel() {
+		// Optimizations do not apply since scoping uses different mechanism
+		// TODO: implement ambient aware scoping
+		return req, true
 	}
 
-	return false
+	req = filterRelevantUpdates(proxy, req)
```

Comment:
Hmm. I do wonder if we may be better served moving this filtering up to this level: https://github.com/istio/istio/blob/91bcb0eb045d30c843817d0c2aba80a12a6531f2/pilot/pkg/xds/ads.go#L591. Doesn't need to block the PR.

Maybe its a bit harder due to the ServiceTargets aspect

Reply:
You mean performing the filtering before enqueuing?

It could be useful to avoid enqueuing altogether, but on the other side, the filtering would be done in a single threaded fashion, while currently it's being done concurrently (limited by push concurrency)